error-prone-contrib/src/main/java/tech/picnic/errorprone/bugpatterns/TypeMemberOrder.java

```diff
@@ -0,0 +1,265 @@
+import java.util.Set;
+import javax.lang.model.element.Modifier;
+
+/**
+ * A {@link BugChecker} that flags classes with a non-canonical member order.
+ *
+ * <p>Class members should be ordered as follows:
+ *
+ * <ol>
+ *   <li>Static fields
+ *   <li>Instance fields
+ *   <li>Static initializer blocks
+ *   <li>Instance initializer blocks
+ *   <li>Constructors
+ *   <li>Methods
+ *   <li>Nested classes, interfaces and enums
+ * </ol>
+ *
+ * @see <a
+ *     href="https://checkstyle.sourceforge.io/apidocs/com/puppycrawl/tools/checkstyle/checks/coding/DeclarationOrderCheck.html">Checkstyle's
+ *     {@code DeclarationOrderCheck}</a>
+ */
+// XXX: Reference
+// https://checkstyle.sourceforge.io/apidocs/com/puppycrawl/tools/checkstyle/checks/coding/DeclarationOrderCheck.html
+@AutoService(BugChecker.class)
+@BugPattern(
+    summary = "Type members should be ordered in a standard way",
+    link = BUG_PATTERNS_BASE_URL + "TypeMemberOrder",
+    linkType = CUSTOM,
+    severity = WARNING,
+    tags = STYLE)
+public final class TypeMemberOrder extends BugChecker implements ClassTreeMatcher {
+  private static final long serialVersionUID = 1L;
+
+  /** Instantiates a new {@link TypeMemberOrder} instance. */
+  public TypeMemberOrder() {}
+
+  @Override
+  public Description matchClass(ClassTree tree, VisitorState state) {
+    Kind kind = tree.getKind();
+    if (kind != Kind.CLASS && kind != Kind.INTERFACE && kind != Kind.ENUM) {
+      return Description.NO_MATCH;
+    }
+
+    // All members that can be moved or may lay between movable ones.
+    ImmutableList<TypeMember> members =
+        tree.getMembers().stream()
+            .filter(member -> isGeneratedConstructor(member) && !isEnumerator(member))
+            .map(m -> new AutoValue_TypeMemberOrder_TypeMember(m, getPreferredOrdinal(m, state)))
+            .collect(toImmutableList());
+
+    // List of the sortable members' preferred ordinals, ordered by the member's position in the
+    // original source.
+    ImmutableList<Integer> preferredOrdinals =
+        members.stream()
+            .filter(m -> m.preferredOrdinal().isPresent())
+            .map(m -> m.preferredOrdinal().orElseThrow(/* Unreachable due to preceding check. */ ))
+            .collect(toImmutableList());
+
+    if (Comparators.isInOrder(preferredOrdinals, naturalOrder())) {
+      return Description.NO_MATCH;
+    }
+
+    int bodyStartPos = getBodyStartPos(tree, state);
+    if (bodyStartPos == Position.NOPOS) {
+      /*
+       * We can't determine the type body's start position in the source code. This generally means
+       * that (part of) its code was generated. Even if the source code for a subset of its members
+       * is available, dealing with this edge case is not worth the trouble.
+       */
+      return Description.NO_MATCH;
+    }
+
+    return describeMatch(tree, sortTypeMembers(bodyStartPos, members, state));
+  }
+
+  private static boolean isGeneratedConstructor(Tree tree) {
```

Comment:
We can maybe move this to `MoreAstHelpers` 🤔 , although it's used only once right now..

WDYT of this implementation:
```
  public static boolean isGeneratedConstructor(Tree tree) {
    return tree instanceof MethodTree methodTree && ASTHelpers.isGeneratedConstructor(methodTree);
  }
```

Reply:
It's inversed, it returns true if `Tree` is  not of type `Kind.METHOD`, will push a commit to fix :).